File: taurus-protect-sdk-python/taurus_protect/mappers/business_rule.py
"""Business rule mapper for converting OpenAPI DTOs to domain models."""

from __future__ import annotations

from typing import Any, List, Optional

from taurus_protect.mappers._base import safe_int, safe_string
from taurus_protect.mappers.currency import currency_from_dto
from taurus_protect.models.business_rule import BusinessRule


def _resolve(dto: Any, snake_name: str, camel_name: str) -> Any:
    """Resolve a field value trying snake_case first, then camelCase."""
    val = getattr(dto, snake_name, None)
    if val is not None:
        return val
    return getattr(dto, camel_name, None)


def business_rule_from_dto(dto: Any) -> Optional[BusinessRule]:
    """Convert OpenAPI TgvalidatordBusinessRule DTO to domain BusinessRule.

    Maps tenantId string -> int (matching Java BusinessRuleMapper).
    """
    if dto is None:
        return None

    raw_id = getattr(dto, "id", None)
    return BusinessRule(
        id=str(raw_id) if raw_id is not None else "",
        tenant_id=safe_int(_resolve(dto, "tenant_id", "tenantId")),
        currency=getattr(dto, "currency", None),
        wallet_id=_resolve(dto, "wallet_id", "walletId"),
        address_id=_resolve(dto, "address_id", "addressId"),
        rule_key=_resolve(dto, "rule_key", "ruleKey"),
        rule_value=_resolve(dto, "rule_value", "ruleValue"),
        rule_group=_resolve(dto, "rule_group", "ruleGroup"),
        rule_description=_resolve(dto, "rule_description", "ruleDescription"),
        rule_validation=_resolve(dto, "rule_validation", "ruleValidation"),
        entity_type=_resolve(dto, "entity_type", "entityType"),
        entity_id=_resolve(dto, "entity_id", "entityID"),
        currency_info=currency_from_dto(_resolve(dto, "currency_info", "currencyInfo")),
    )


def business_rules_from_dto(dtos: Optional[List[Any]]) -> List[BusinessRule]:
    """Convert list of OpenAPI business rule DTOs to domain BusinessRules."""
    if dtos is None:
        return []
    return [r for dto in dtos if (r := business_rule_from_dto(dto)) is not None]

File: taurus-protect-sdk-go/pkg/protect/mapper/business_rule.go
package mapper

import (
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/internal/openapi"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// BusinessRuleFromDTO converts an OpenAPI TgvalidatordBusinessRule to a domain BusinessRule.
func BusinessRuleFromDTO(dto *openapi.TgvalidatordBusinessRule) *model.BusinessRule {
	if dto == nil {
		return nil
	}

	rule := &model.BusinessRule{
		ID:              safeString(dto.Id),
		TenantID:        safeString(dto.TenantId),
		Currency:        safeString(dto.Currency),
		WalletID:        safeString(dto.WalletId),
		RuleKey:         safeString(dto.RuleKey),
		RuleValue:       safeString(dto.RuleValue),
		RuleGroup:       safeString(dto.RuleGroup),
		RuleDescription: safeString(dto.RuleDescription),
		RuleValidation:  safeString(dto.RuleValidation),
		AddressID:       safeString(dto.AddressId),
		EntityType:      safeString(dto.EntityType),
		EntityID:        safeString(dto.EntityID),
	}

	// Convert currency info if present
	if dto.CurrencyInfo != nil {
		rule.CurrencyInfo = CurrencyFromDTO(dto.CurrencyInfo)
	}

	return rule
}

// BusinessRulesFromDTO converts a slice of OpenAPI TgvalidatordBusinessRule to domain BusinessRules.
func BusinessRulesFromDTO(dtos []openapi.TgvalidatordBusinessRule) []*model.BusinessRule {
	if dtos == nil {
		return nil
	}
	rules := make([]*model.BusinessRule, len(dtos))
	for i := range dtos {
		rules[i] = BusinessRuleFromDTO(&dtos[i])
	}
	return rules
}
